File: src/vs/workbench/contrib/computerUse/helper/darwin/Sources/v3code-computer-use/Observation.swift
/*--------------------------------------------------------------------------------------
 *  Copyright 2025 V3Code / Glass Devtools. All rights reserved.
 *  Licensed under the Apache License, Version 2.0.
 *--------------------------------------------------------------------------------------*/

import CoreGraphics
import Foundation

/// Ambient observation: sampling one application on a timer rather than because a tool call asked.
///
/// This is the most privileged thing in the helper — it observes with no tool call behind it — so the
/// entire design is arranged so that no single mistake can turn it on, mirroring
/// `common/computerUseObservation.ts` on the other side of the pipe.
///
/// **Nothing is sampled until `observeStart` succeeds.** There is no default session, no implicit
/// start, no "observe the frontmost application" convenience, and no code path from any other method
/// into `sample`. The timer does not exist until a session does, and it is torn down with the last one.
///
/// **Every check refuses.** `start` is a chain of guards ending in one success. A pid that cannot be
/// identified, an application whose tier is `self`, an `appId` that does not match the pid, an interval
/// outside the permitted band, a `stopAtMs` in the past, a screenshot session without Screen Recording
/// — each is a refusal, and there is no fallback value for any of them. The same guards run again on
/// every tick, so a grant that stops being true stops the session rather than waiting for someone to
/// notice.
///
/// **`stopAtMs` is the whole safety story and is enforced twice.** The caller sets it from the
/// `permittedUntil` of the policy decision that authorised the session. It is clamped here to
/// `observationMaxGrantMs` from now, checked by the supervisor tick, and checked again inside the
/// sample itself. If V3Code crashes, is force-quit, or loses the pipe, the session expires on its own;
/// and if the pipe closes the dispatcher stops everything outright, so a lost `observeStop` cannot
/// leave capture running.
///
/// **Why the helper duplicates the policy engine's arithmetic.** The service already decided; this
/// layer exists because the helper is the last thing between a model and the user's screen. A bug in
/// the service, a stale renderer, or anything that learns to speak this protocol must still be
/// refused. The duplication is deliberate, exactly as `AppTiers` duplicates the tier fragments.
///
/// **What a sample is, and the gap this phase leaves.** The protocol has no channel for delivering a
/// sample: all three `observe*` methods return only `ComputerUseObserveStatusResult`, whose sole
/// per-session payload is a count. So a sample here refreshes what the helper alone can hold — the
/// stable refs and the change watermark for that application, which is what makes a later `axTreeDiff`
/// answer `unchanged` in one round-trip — and, for a screenshot session, digests the window frames to
/// learn whether the screen moved. **Pixels are digested and dropped inside `FrameSampler`; no image is
/// encoded, retained, or written anywhere by this file.** Retaining screen images in a process with no
/// consumer for them and no retention owner is precisely what the policy engine's comments warn
/// against, so it is not done. Delivering samples needs a wire shape that does not exist yet.
final class Observation {
	// ---------------------------------------------------------------------------------------------
	// Bounds
	// ---------------------------------------------------------------------------------------------

	/// Mirrors `COMPUTER_USE_OBSERVATION_MAX_GRANT_MS`. A session may never outlive this from its start.
	static let maxGrantMs: Int64 = 8 * 60 * 60 * 1000

	/// Floor on the sampling interval.
	///
	/// Twice a second is the boundary between *ambient observation* and *recording*. A caller asking for
	/// faster is either mistaken or is trying to use this as a screen recorder, and both are refused
	/// rather than clamped: silently sampling slower than asked would leave the caller believing it had
	/// something it does not.
	static let minIntervalMs = 500

	/// Ceiling on the sampling interval. An hour between samples is not observation, it is a leak of a
	/// session nobody remembers starting.
	static let maxIntervalMs = 60 * 60 * 1000

	/// How many applications may be observed at once, so the cost stays bounded and reviewable.
	static let maxSessions = 4

	/// How often the supervisor wakes. Fine enough that `stopAtMs` is honoured to within a quarter
	/// second, coarse enough to be free when nothing is due.
	static let supervisorIntervalMs = 250

	// ---------------------------------------------------------------------------------------------
	// State
	// ---------------------------------------------------------------------------------------------

	private final class Session {
		let pid: pid_t
		let appId: String
		let content: ObserveContent
		let intervalMs: Int
		let startedAt: Int64
		let stopAtMs: Int64
		/// Carried so the session records what it was authorised for, even though nothing here produces an
		/// image to bound. See the delivery note on the type comment.
		let maxLongEdge: Int?
		var samples = 0
		/// Samples whose frame digest differed from the previous one. Logged, not yet on the wire.
		var frameChanges = 0
		var lastFrameHash: UInt64?
		var nextDueAt: Date
		/// True while a sample is on the work queue, so a slow walk cannot pile up behind itself.
		var sampling = false

		init(
			pid: pid_t,
			appId: String,
			content: ObserveContent,
			intervalMs: Int,
			startedAt: Int64,
			stopAtMs: Int64,
			maxLongEdge: Int?
		) {
			self.pid = pid
			self.appId = appId
			self.content = content
			self.intervalMs = intervalMs
			self.startedAt = startedAt
			self.stopAtMs = stopAtMs
			self.maxLongEdge = maxLongEdge
			// Due immediately: the first sample is the one the caller is waiting for, and delaying it by a
			// whole interval would make a long-interval session look wedged.
			self.nextDueAt = Date()
		}
	}

	private let lock = NSLock()
	private var sessions: [pid_t: Session] = [:]
	private var supervisor: DispatchSourceTimer?

	private let workQueue: DispatchQueue
	private let timerQueue = DispatchQueue(label: "dev.v3code.computerUse.observe", qos: .utility)
	private let refTable: RefTable
	private let snapshots: AxSnapshots
	private let watcher: AxWatcher

	init(workQueue: DispatchQueue, refTable: RefTable, snapshots: AxSnapshots, watcher: AxWatcher) {
		self.workQueue = workQueue
		self.refTable = refTable
		self.snapshots = snapshots
		self.watcher = watcher
	}

	// ---------------------------------------------------------------------------------------------
	// Start
	// ---------------------------------------------------------------------------------------------

	/// Validates and starts a session, or throws. Never partially starts.
	func start(_ params: ObserveStartParams) throws -> ObserveStatusResult {
		guard Ax.isProcessTrusted() else {
			throw HelperError(
				.accessibilityNotTrusted,
				"Accessibility permission has not been granted to the computer-use helper"
			)
		}

		let pid = params.pid
		guard let identity = Apps.identity(pid: pid) else {
			throw HelperError(
				.targetNotFound,
				"no running application with pid \(pid), so there is nothing to observe"
			)
		}

		// Deliberately no policy check here. Whether an application may be observed is the user's decision,
		// recorded as a grant the renderer service holds and enforces; the helper cannot see that grant, so
		// anything it decided here could only contradict it. Observing V3Code itself is a poor idea — the
		// agent reads its own output back — but it is discouraged at the approval prompt, not forbidden.
		//
		// What this method DOES enforce is that the authorisation actually matches what would be observed,
		// and that it expires. Those are not policy; they are the difference between an authorised session
		// and an unbounded one.

		// The policy decision the service made names an application; the request names a pid. If the two
		// disagree, the authorisation belongs to something other than what would be observed, so the
		// request is refused rather than reconciled.
		let requested = Observation.normalize(params.appId)
		let actual = Observation.normalize(identity.id)
		guard !requested.isEmpty, requested == actual else {
			throw HelperError(
				.permissionDenied,
				"pid \(pid) is '\(identity.id)', not '\(params.appId)'; the observation grant does not cover it",
				retryable: false
			)
		}

		guard params.intervalMs >= Observation.minIntervalMs, params.intervalMs <= Observation.maxIntervalMs else {
			throw HelperError(
				.permissionDenied,
				"an observation interval of \(params.intervalMs)ms is outside the permitted "
					+ "\(Observation.minIntervalMs)-\(Observation.maxIntervalMs)ms band",
				retryable: false
			)
		}

		let now = Observation.nowMs()
		guard params.stopAtMs > now else {
			throw HelperError(
				.permissionDenied,
				"the observation grant expires at \(params.stopAtMs), which is not in the future",
				retryable: false
			)
		}
		// Clamped rather than refused: an over-long grant is more likely a mistake than an attack, and
		// shortening it is the fail-closed direction.
		let stopAtMs = min(params.stopAtMs, now + Observation.maxGrantMs)
		if stopAtMs != params.stopAtMs {
			Log.warn("clamped an observation grant for pid \(pid) from \(params.stopAtMs) to \(stopAtMs)")
		}

		if params.content == .axTreeAndScreenshots {
			// Refused at start rather than discovered per sample: a session that silently never captures
			// the thing it was authorised for is worse than one that never begins.
			guard CGPreflightScreenCaptureAccess() else {
				throw HelperError(
					.screenRecordingNotGranted,
					"Screen Recording permission has not been granted to the computer-use helper, so an "
						+ "observation session including screenshots cannot start"
				)
			}
		}

		lock.lock()
		if sessions[pid] == nil, sessions.count >= Observation.maxSessions {
			lock.unlock()
			throw HelperError(
				.permissionDenied,
				"already observing \(Observation.maxSessions) applications, which is the limit",
				retryable: true
			)
		}
		// Starting again for the same pid replaces the session outright, so a re-authorisation cannot end
		// up with two timers sampling the same application.
		let session = Session(
			pid: pid,
			appId: actual,
			content: params.content,
			intervalMs: params.intervalMs,
			startedAt: now,
			stopAtMs: stopAtMs,
			maxLongEdge: params.maxLongEdge
		)
		sessions[pid] = session
		let snapshot = describeLocked()
		lock.unlock()

		watcher.ensureWatching(pid: pid)
		startSupervisor()
		Log.info(
			"observing \(identity.id) (pid \(pid)) every \(params.intervalMs)ms until \(stopAtMs), content \(params.content.rawValue)"
		)
		return ObserveStatusResult(sessions: snapshot)
	}

	// ---------------------------------------------------------------------------------------------
	// Stop
	// ---------------------------------------------------------------------------------------------

	/// Stops one session, or every session when `pid` is nil. Always succeeds.
	///
	/// Never throws and never gates: stopping is the safe direction, and a revocation that could be
	/// refused would be a revocation that does not work.
	@discardableResult
	func stop(pid: pid_t?) -> ObserveStatusResult {
		lock.lock()
		if let pid {
			if sessions.removeValue(forKey: pid) != nil {
				Log.info("stopped observing pid \(pid)")
			}
		} else if !sessions.isEmpty {
			Log.info("stopped observing every application (\(sessions.count) sessions)")
			sessions.removeAll()
		}
		let empty = sessions.isEmpty
		let snapshot = describeLocked()
		lock.unlock()

		if empty {
			stopSupervisor()
		}
		return ObserveStatusResult(sessions: snapshot)
	}

	/// What the helper is currently observing.
	func status() -> ObserveStatusResult {
		lock.lock()
		defer { lock.unlock() }
		return ObserveStatusResult(sessions: describeLocked())
	}

	// ---------------------------------------------------------------------------------------------
	// Supervisor
	// ---------------------------------------------------------------------------------------------

	private func startSupervisor() {
		lock.lock()
		defer { lock.unlock() }
		guard supervisor == nil else { return }
		let timer = DispatchSource.makeTimerSource(queue: timerQueue)
		timer.schedule(
			deadline: .now(),
			repeating: .milliseconds(Observation.supervisorIntervalMs),
			leeway: .milliseconds(50)
		)
		timer.setEventHandler { [weak self] in
			self?.tick()
		}
		supervisor = timer
		timer.resume()
	}

	private func stopSupervisor() {
		lock.lock()
		let timer = supervisor
		supervisor = nil
		lock.unlock()
		timer?.cancel()
	}

	/// One supervisor pass: expire what is over, dispatch what is due.
	private func tick() {
		let now = Observation.nowMs()

		lock.lock()
		var expired: [pid_t] = []
		var due: [Session] = []
		let wallClock = Date()
		for session in sessions.values {
			if now >= session.stopAtMs {
				expired.append(session.pid)
				continue
			}
			if !session.sampling, wallClock >= session.nextDueAt {
				session.sampling = true
				due.append(session)
			}
		}
		for pid in expired {
			sessions.removeValue(forKey: pid)
		}
		let empty = sessions.isEmpty
		lock.unlock()

		for pid in expired {
			Log.info("observation of pid \(pid) reached its grant expiry and stopped itself")
		}
		if empty {
			stopSupervisor()
		}

		Log.debug("observation tick: \(due.count) due, \(expired.count) expired")
		for session in due {
			// Samples run on the shared work queue, not here: the queue is serial precisely so a tree walk
			// cannot interleave with an action, and an ambient sample is no more entitled to break that
			// than a tool call is.
			workQueue.async { [weak self] in
				self?.sample(session)
			}
		}
	}

	/// Takes one sample, re-checking every reason not to first.
	private func sample(_ session: Session) {
		defer {
			lock.lock()
			session.sampling = false
			session.nextDueAt = Date().addingTimeInterval(Double(session.intervalMs) / 1000)
			lock.unlock()
		}

		// Re-validated on every tick rather than trusted from start time. A session outlives the moment
		// it was authorised, and each of these can stop being true while it runs.
		guard Observation.nowMs() < session.stopAtMs else {
			stop(pid: session.pid)
			return
		}
		guard Ax.isProcessTrusted() else {
			Log.warn("Accessibility trust was lost; stopping observation of pid \(session.pid)")
			stop(pid: session.pid)
			return
		}
		guard let identity = Apps.identity(pid: session.pid) else {
			Log.info("pid \(session.pid) is gone; stopping its observation")
			stop(pid: session.pid)
			snapshots.forget(pid: session.pid)
			watcher.stopWatching(pid: session.pid)
			return
		}
		// Identity only — not policy. If the pid now belongs to a different application than the grant
		// covered, the authorisation does not apply to what is on screen and sampling must stop.
		guard Observation.normalize(identity.id) == session.appId else {
			// The pid now belongs to a different application than the grant covered — pid reuse. Stop.
			Log.warn("pid \(session.pid) is now '\(identity.id)'; stopping observation authorised for '\(session.appId)'")
			stop(pid: session.pid)
			return
		}
		if session.content == .axTreeAndScreenshots, !CGPreflightScreenCaptureAccess() {
			Log.warn("Screen Recording was revoked; stopping observation of pid \(session.pid)")
			stop(pid: session.pid)
			return
		}

		let token = Cancellation.Token.uncancellable
		do {
			// Walk and fingerprint inside one suppression scope, counter outside it, for the reason set out
			// on `Dispatcher.walk`: reading provokes notifications, and a sample that counted its own
			// footprints would make every later `axTreeDiff` believe the application had done something.
			let (tree, fingerprint) = try watcher.withSuppression(pid: session.pid) {
				let tree = try AxTreeReader.read(
					pid: session.pid,
					maxDepth: nil,
					refTable: refTable,
					cancellation: token
				)
				return (tree, AxSnapshots.fingerprint(pid: session.pid))
			}
			let activity = watcher.activity(pid: session.pid)
			snapshots.record(
				pid: session.pid,
				generation: tree.generation,
				activity: activity.count,
				notificationsAvailable: activity.notificationsAvailable,
				fingerprint: fingerprint
			)

			if session.content == .axTreeAndScreenshots {
				// Digest only. See the type comment: no image is encoded or kept.
				if let hash = FrameSampler.fingerprint(pid: session.pid) {
					lock.lock()
					if let previous = session.lastFrameHash, previous != hash {
						session.frameChanges += 1
					}
					session.lastFrameHash = hash
					lock.unlock()
				}
			}

			lock.lock()
			session.samples += 1
			let count = session.samples
			let changes = session.frameChanges
			lock.unlock()
			Log.debug(
				"observation sample \(count) of \(session.appId): generation \(tree.generation), \(tree.nodes.count) roots, \(changes) frame changes"
			)
		} catch let error as HelperError {
			// A failing sample stops the session rather than retrying forever: the two things that make a
			// walk fail here — the application died, or permission went away — are both permanent, and a
			// session that logs an error every interval is a session nobody will read the logs of.
			Log.warn("observation of pid \(session.pid) failed (\(error.code.rawValue): \(error.message)); stopping")
			stop(pid: session.pid)
		} catch {
			Log.warn("observation of pid \(session.pid) failed unexpectedly (\(error)); stopping")
			stop(pid: session.pid)
		}
	}

	// ---------------------------------------------------------------------------------------------
	// Helpers
	// ---------------------------------------------------------------------------------------------

	/// Wire shapes for the current sessions. Caller must hold the lock.
	private func describeLocked() -> [ObserveSession] {
		sessions.values
			.map { session in
				ObserveSession(
					pid: session.pid,
					appId: session.appId,
					startedAt: session.startedAt,
					stopAtMs: session.stopAtMs,
					intervalMs: session.intervalMs,
					content: session.content,
					samples: session.samples
				)
			}
			.sorted { $0.pid < $1.pid }
	}

	/// Matches `normalizeAppId` in computerUseObservation.ts: case and surrounding whitespace only.
	///
	/// Nothing fuzzy, deliberately. Fragment matching is right for tier classification, where
	/// over-matching is restrictive; here it would be permissive, and a grant for one application would
	/// quietly cover another that shares a prefix.
	private static func normalize(_ id: String) -> String {
		id.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
	}

	static func nowMs() -> Int64 {
		Int64((Date().timeIntervalSince1970 * 1000).rounded())
	}
}

File: src/vs/workbench/contrib/computerUse/helper/win32/Observation.cpp
/*--------------------------------------------------------------------------------------
 *  Copyright 2025 V3Code / Glass Devtools. All rights reserved.
 *  Licensed under the Apache License, Version 2.0.
 *--------------------------------------------------------------------------------------*/
//
// Ambient observation implementation. Read Observation.h FIRST. The four rules there — nothing before
// start, stopAtMs enforced here, never self, everything the caller supplies is clamped — are the design,
// and the note about what a sample deliberately is not is the part most likely to be misread.
//
// NEEDS VERIFICATION ON WINDOWS (full list in Observation.h):
//   - Self-expiry after V3Code is killed. That is the test that matters.
//   - That the sampling thread's own IUIAutomation instance works and never touches UiaClient's.
//   - That Shutdown() joins cleanly on stdin EOF with no orphan left behind.
//
#include "Observation.h"

#include "AppTiers.h"
#include "Log.h"
#include "UiaClient.h"

#include <chrono>
#include <condition_variable>
#include <cstddef>
#include <mutex>
#include <thread>
#include <vector>

#include <uiautomation.h>
#include <wrl/client.h>

namespace v3cu {
namespace {

using Microsoft::WRL::ComPtr;

/// One running session, as ComputerUseObserveSession describes it.
struct Session {
	unsigned long pid = 0;
	std::string appId;
	int intervalMs = 1000;
	int64_t startedAt = 0;
	int64_t stopAtMs = 0;
	bool screenshots = false;
	uint64_t samples = 0;
	/// GetTickCount64 value of the next due sample. Ticks rather than epoch, so a system clock change
	/// cannot make a session sample in a tight loop.
	ULONGLONG nextDueTick = 0;
};

int64_t NowEpochMs() {
	// Epoch milliseconds, matching Date.now() on the other side of the pipe. stopAtMs is a wall-clock
	// instant negotiated with the policy engine, so it has to be compared against wall clock even though
	// scheduling uses ticks.
	const auto now = std::chrono::system_clock::now().time_since_epoch();
	return std::chrono::duration_cast<std::chrono::milliseconds>(now).count();
}

/// Process-wide observation state. A namespace-scope singleton rather than members on Observation so the
/// header does not have to expose a mutex, a thread and a condition variable to every file that includes
/// it.
struct State {
	std::mutex mutex;
	std::condition_variable wake;
	std::vector<Session> sessions;
	std::thread thread;
	bool running = false;
	bool stopping = false;
};

State& Shared() {
	static State state;
	return state;
}

/// Drops sessions whose stop time has passed. Caller holds the lock.
void PruneExpiredLocked(std::vector<Session>& sessions) {
	const int64_t now = NowEpochMs();
	for (size_t index = sessions.size(); index > 0; --index) {
		Session& session = sessions[index - 1];
		if (session.stopAtMs > now) {
			continue;
		}
		LogInfo("observation of " + session.appId + " (pid " + std::to_string(session.pid) +
			") expired after " + std::to_string(session.samples) + " samples");
		sessions.erase(sessions.begin() + static_cast<ptrdiff_t>(index - 1));
	}
}

JsonValue SerializeSession(const Session& session) {
	JsonValue json = JsonValue::MakeObject();
	json.Set("pid", JsonValue::MakeInt(static_cast<int64_t>(session.pid)));
	json.Set("appId", JsonValue::MakeString(session.appId));
	json.Set("startedAt", JsonValue::MakeInt(session.startedAt));
	json.Set("stopAtMs", JsonValue::MakeInt(session.stopAtMs));
	json.Set("intervalMs", JsonValue::MakeInt(session.intervalMs));
	json.Set("content",
		JsonValue::MakeString(session.screenshots ? "axTreeAndScreenshots" : "axTree"));
	json.Set("samples", JsonValue::MakeInt(static_cast<int64_t>(session.samples)));
	return json;
}

JsonValue SerializeStatusLocked(const std::vector<Session>& sessions) {
	JsonValue list = JsonValue::MakeArray();
	for (const Session& session : sessions) {
		list.Push(SerializeSession(session));
	}
	JsonValue result = JsonValue::MakeObject();
	result.Set("observing", JsonValue::MakeBool(!sessions.empty()));
	result.Set("sessions", std::move(list));
	return result;
}

/// Counts children breadth-first up to a cap. Deliberately shallow: a sample exists to prove the session
/// can still observe the application, not to produce a tree — nothing can consume a tree (Observation.h).
int CountSampleNodes(IUIAutomation* automation, IUIAutomationTreeWalker* walker, HWND window) {
	if (automation == nullptr || walker == nullptr || window == nullptr) {
		return 0;
	}
	ComPtr<IUIAutomationElement> root;
	if (FAILED(automation->ElementFromHandle(window, &root)) || !root) {
		return 0;
	}

	int counted = 1;
	std::vector<ComPtr<IUIAutomationElement>> frontier;
	frontier.push_back(root);
	while (!frontier.empty() && counted < kMaxObserveSampleNodes) {
		ComPtr<IUIAutomationElement> current = frontier.back();
		frontier.pop_back();

		ComPtr<IUIAutomationElement> child;
		if (FAILED(walker->GetFirstChildElement(current.Get(), &child)) || !child) {
			continue;
		}
		while (child && counted < kMaxObserveSampleNodes) {
			++counted;
			frontier.push_back(child);
			ComPtr<IUIAutomationElement> sibling;
			if (FAILED(walker->GetNextSiblingElement(child.Get(), &sibling)) || !sibling) {
				break;
			}
			child = sibling;
		}
	}
	return counted;
}

void SampleLoop() {
	// This thread's own apartment and its own automation object. It must not use UiaClient::Instance(),
	// which belongs to the worker thread — see the note in Observation.h.
	ComApartment apartment;
	if (!apartment.Ok()) {
		LogError("observation thread could not initialize COM; sessions will not sample");
	}

	ComPtr<IUIAutomation> automation;
	ComPtr<IUIAutomationTreeWalker> walker;
	if (apartment.Ok()) {
		if (FAILED(::CoCreateInstance(CLSID_CUIAutomation, nullptr, CLSCTX_INPROC_SERVER,
				IID_PPV_ARGS(&automation))) ||
			!automation) {
			LogWarn("observation thread could not create a UI Automation client");
		} else if (FAILED(automation->get_ControlViewWalker(&walker))) {
			walker.Reset();
		}
	}

	State& state = Shared();
	for (;;) {
		std::vector<unsigned long> due;
		{
			std::unique_lock<std::mutex> guard(state.mutex);
			if (state.stopping) {
				return;
			}

			PruneExpiredLocked(state.sessions);
			if (state.sessions.empty()) {
				// Nothing to do and nothing scheduled. Wait to be woken by a Start or by Shutdown rather
				// than spinning on an empty list.
				state.wake.wait(guard, [&state]() { return state.stopping || !state.sessions.empty(); });
				if (state.stopping) {
					return;
				}
				continue;
			}

			const ULONGLONG now = ::GetTickCount64();
			ULONGLONG nextDue = now + static_cast<ULONGLONG>(kMaxObserveIntervalMs);
			for (Session& session : state.sessions) {
				if (session.nextDueTick <= now) {
					due.push_back(session.pid);
					session.nextDueTick = now + static_cast<ULONGLONG>(session.intervalMs);
				}
				if (session.nextDueTick < nextDue) {
					nextDue = session.nextDueTick;
				}
			}

			if (due.empty()) {
				const ULONGLONG waitMs = nextDue > now ? nextDue - now : 1;
				state.wake.wait_for(guard, std::chrono::milliseconds(waitMs));
				continue;
			}
		}

		// Sampled with the lock released: a UIA round-trip into a hung application must not block
		// observeStop.
		for (const unsigned long pid : due) {
			const HWND window = FindMainWindowForPid(pid);
			if (window == nullptr) {
				// The application closed its windows. Not an error and not a reason to stop the session:
				// stopAtMs and the caller decide when a session ends, not a transiently window-less app.
				continue;
			}
			const int nodes = CountSampleNodes(automation.Get(), walker.Get(), window);

			// >>> THIS IS WHERE THE NEXT PHASE HOOKS IN. When the protocol grows a delivery path, the
			// sample's content is produced here — the accessibility tree, and a capture for a session whose
			// content includes screenshots — and handed to that path. Until then nothing is produced and
			// nothing is retained, deliberately; see Observation.h. A `0 nodes` line here every interval
			// means this thread has no working UI Automation client, which is worth knowing.
			LogDebug("observation sample: pid " + std::to_string(pid) + ", " + std::to_string(nodes) +
				" nodes visible");

			std::lock_guard<std::mutex> guard(state.mutex);
			for (Session& session : state.sessions) {
				if (session.pid == pid) {
					// Counts ATTEMPTS, which is what the contract needs it for: telling a live session from a
					// wedged one. A sample that read nothing still proves the timer is running, and the debug
					// line above is what says whether it read anything.
					++session.samples;
					break;
				}
			}
		}
	}
}

} // namespace

Observation& Observation::Instance() {
	static Observation observation;
	return observation;
}

bool Observation::Start(const ObserveStartOptions& options, ErrorCode& code, std::string& message) {
	if (options.pid == 0) {
		code = ErrorCode::Internal;
		message = "observeStart requires a pid";
		return false;
	}
	if (options.pid == static_cast<unsigned long>(::GetCurrentProcessId())) {
		// Not policy: the helper is a background process with no window, so observing it samples nothing
		// forever. Refusing is a correctness answer, not a permission one.
		code = ErrorCode::TargetNotFound;
		message = "the helper has no window to observe";
		return false;
	}

	AppInfo app;
	if (!GetAppInfoForPid(options.pid, app)) {
		code = ErrorCode::TargetNotFound;
		message = "no process with pid " + std::to_string(options.pid);
		return false;
	}
	// No check for "is this V3Code". Observing V3Code's own window is a poor idea — the agent reads its
	// own output back — but it is discouraged at the approval prompt rather than forbidden here, because
	// the grant is the user's to make. Mirrors helper/darwin Observation.swift.

	if (FindMainWindowForPid(options.pid) == nullptr) {
		code = ErrorCode::TargetNotFound;
		message = "application " + app.id + " has no visible top-level window";
		return false;
	}

	const int64_t now = NowEpochMs();
	if (options.stopAtMs <= now) {
		// Mandatory and load-bearing: a session with no future stop time is a session nothing can be
		// relied upon to end.
		code = ErrorCode::Internal;
		message = "observeStart requires a stopAtMs in the future";
		return false;
	}

	Session session;
	session.pid = options.pid;
	session.appId = options.appId.empty() ? app.id : options.appId;
	session.intervalMs = Clamp(options.intervalMs, kMinObserveIntervalMs, kMaxObserveIntervalMs);
	session.startedAt = now;
	// Clamped, not trusted. Rule 4.
	session.stopAtMs = options.stopAtMs > now + kMaxObserveGrantMs ? now + kMaxObserveGrantMs
																  : options.stopAtMs;
	session.screenshots = options.screenshots;
	session.nextDueTick = ::GetTickCount64();

	if (session.screenshots) {
		// Said out loud, in the log the user's bug report will contain, rather than only in a header
		// comment: the request is recorded but screenshots are not taken. See Observation.h.
		LogWarn("observeStart asked for screenshots of " + session.appId +
			"; protocol 2 has no way to deliver a sample, so no screenshot is captured");
	}
	if (!options.excludePids.empty() || options.maxLongEdge > 0) {
		LogDebug("observeStart screenshot options ignored: no screenshot is captured");
	}

	State& state = Shared();
	{
		std::lock_guard<std::mutex> guard(state.mutex);
		PruneExpiredLocked(state.sessions);

		bool replaced = false;
		for (Session& existing : state.sessions) {
			if (existing.pid == session.pid) {
				// One session per application. Restarting is how a caller extends or narrows a grant, and
				// two sessions for one pid would sample it twice and report it twice.
				existing = session;
				replaced = true;
				break;
			}
		}
		if (!replaced) {
			if (state.sessions.size() >= kMaxObserveSessions) {
				code = ErrorCode::PermissionDenied;
				message = "at most " + std::to_string(kMaxObserveSessions) +
					" applications may be observed at once";
				return false;
			}
			state.sessions.push_back(session);
		}

		if (!state.running) {
			state.running = true;
			state.stopping = false;
			state.thread = std::thread(SampleLoop);
		}
	}
	state.wake.notify_all();

	LogInfo("observing " + session.appId + " (pid " + std::to_string(session.pid) + ") every " +
		std::to_string(session.intervalMs) + "ms until epoch " + std::to_string(session.stopAtMs));
	return true;
}

void Observation::Stop(bool all, unsigned long pid) {
	State& state = Shared();
	{
		std::lock_guard<std::mutex> guard(state.mutex);
		if (all) {
			if (!state.sessions.empty()) {
				LogInfo("stopping all " + std::to_string(state.sessions.size()) +
					" observation session(s)");
			}
			state.sessions.clear();
		} else {
			for (size_t index = 0; index < state.sessions.size(); ++index) {
				if (state.sessions[index].pid == pid) {
					LogInfo("stopping observation of " + state.sessions[index].appId);
					state.sessions.erase(state.sessions.begin() + static_cast<ptrdiff_t>(index));
					break;
				}
			}
		}
	}
	state.wake.notify_all();
}

JsonValue Observation::Status() {
	State& state = Shared();
	std::lock_guard<std::mutex> guard(state.mutex);
	// Pruned before reporting, so an expired session can never be reported as running — not even if the
	// sampling thread is stuck inside a UIA call. observeStatus is the reconciliation input for the
	// service, so it has to be true at the moment it is asked, not eventually.
	PruneExpiredLocked(state.sessions);
	return SerializeStatusLocked(state.sessions);
}

void Observation::Shutdown() {
	State& state = Shared();
	std::thread thread;
	{
		std::lock_guard<std::mutex> guard(state.mutex);
		state.sessions.clear();
		state.stopping = true;
		thread = std::move(state.thread);
		state.running = false;
	}
	state.wake.notify_all();
	if (thread.joinable()) {
		thread.join();
	}
}

} // namespace v3cu
